{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pyarrow.parquet import ParquetFile\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "11it [25:02, 136.58s/it]\n"
     ]
    }
   ],
   "source": [
    "multiome_var_meta = pd.read_csv(\"../../data/multiome_var_meta.csv\")\n",
    "gene_names = multiome_var_meta[multiome_var_meta[\"feature_type\"] == \"Gene Expression\"][\"location\"].to_list()\n",
    "\n",
    "multiome_df = pd.DataFrame()\n",
    "batch_num = 10\n",
    "\n",
    "def get_chr(all_str):\n",
    "    index = all_str.find(\":\") \n",
    "    return all_str[:index] \n",
    "\n",
    "def get_start(all_str):\n",
    "    return int(all_str[all_str.find(\":\")+1:all_str.find(\"-\")])\n",
    "\n",
    "def get_end(all_str):\n",
    "    return int(all_str[all_str.find(\"-\")+1:])\n",
    "\n",
    "multiome_file = ParquetFile(\"../../data/multiome_train.parquet\")\n",
    "batch_group_size = multiome_file.metadata.num_rows // batch_num\n",
    "for batch in tqdm(multiome_file.iter_batches(batch_size=batch_group_size)):\n",
    "    multiome_batch = batch.to_pandas()\n",
    "    multiome_batch = multiome_batch[~multiome_batch[\"location\"].isin(gene_names)]\n",
    "\n",
    "    multiome_batch[\"chromosome\"] = multiome_batch[\"location\"].apply(get_chr)\n",
    "    multiome_batch[\"start\"] = multiome_batch[\"location\"].apply(get_start)\n",
    "    multiome_batch[\"end\"] = multiome_batch[\"location\"].apply(get_end)\n",
    "\n",
    "    multiome_df = pd.concat([multiome_df, multiome_batch], axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'multiome_df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\ghanb\\Desktop\\Main\\Projects\\Kaggle_SCP\\notebooks\\notebooks_ temp\\ATACseq_preparation_yalback.ipynb Cell 3\u001b[0m line \u001b[0;36m1\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/ghanb/Desktop/Main/Projects/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation_yalback.ipynb#W2sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m multiome_df\u001b[39m.\u001b[39mto_parquet(\u001b[39m\"\u001b[39m\u001b[39m../../data/ATAC_train.parquet\u001b[39m\u001b[39m\"\u001b[39m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'multiome_df' is not defined"
     ]
    }
   ],
   "source": [
    "multiome_df.to_parquet(\"../../data/ATAC_train.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiome_df = pd.read_parquet(\"../../data/temp/ATAC_train.parquet\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>location</th>\n",
       "      <th>count</th>\n",
       "      <th>normalized_count</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>000225c1151ab841</td>\n",
       "      <td>chr10:100346599-100347467</td>\n",
       "      <td>2</td>\n",
       "      <td>2.664076</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100346599</td>\n",
       "      <td>100347467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>000225c1151ab841</td>\n",
       "      <td>chr10:101214564-101215355</td>\n",
       "      <td>2</td>\n",
       "      <td>4.966185</td>\n",
       "      <td>chr10</td>\n",
       "      <td>101214564</td>\n",
       "      <td>101215355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>000225c1151ab841</td>\n",
       "      <td>chr10:101353529-101354425</td>\n",
       "      <td>4</td>\n",
       "      <td>1.627825</td>\n",
       "      <td>chr10</td>\n",
       "      <td>101353529</td>\n",
       "      <td>101354425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>000225c1151ab841</td>\n",
       "      <td>chr10:101830364-101831303</td>\n",
       "      <td>2</td>\n",
       "      <td>1.333610</td>\n",
       "      <td>chr10</td>\n",
       "      <td>101830364</td>\n",
       "      <td>101831303</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>000225c1151ab841</td>\n",
       "      <td>chr10:102064862-102065750</td>\n",
       "      <td>2</td>\n",
       "      <td>1.588739</td>\n",
       "      <td>chr10</td>\n",
       "      <td>102064862</td>\n",
       "      <td>102065750</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173485197</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:19499847-19500746</td>\n",
       "      <td>1</td>\n",
       "      <td>3.342186</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19499847</td>\n",
       "      <td>19500746</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173485198</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:19566990-19567754</td>\n",
       "      <td>1</td>\n",
       "      <td>1.154666</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19566990</td>\n",
       "      <td>19567754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173485199</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:20735890-20736626</td>\n",
       "      <td>3</td>\n",
       "      <td>7.049924</td>\n",
       "      <td>chrY</td>\n",
       "      <td>20735890</td>\n",
       "      <td>20736626</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173485200</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:21260054-21260842</td>\n",
       "      <td>2</td>\n",
       "      <td>3.967924</td>\n",
       "      <td>chrY</td>\n",
       "      <td>21260054</td>\n",
       "      <td>21260842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>173485201</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:7273431-7274301</td>\n",
       "      <td>2</td>\n",
       "      <td>1.250685</td>\n",
       "      <td>chrY</td>\n",
       "      <td>7273431</td>\n",
       "      <td>7274301</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>173485202 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     obs_id                   location  count  \\\n",
       "0          000225c1151ab841  chr10:100346599-100347467      2   \n",
       "1          000225c1151ab841  chr10:101214564-101215355      2   \n",
       "2          000225c1151ab841  chr10:101353529-101354425      4   \n",
       "3          000225c1151ab841  chr10:101830364-101831303      2   \n",
       "4          000225c1151ab841  chr10:102064862-102065750      2   \n",
       "...                     ...                        ...    ...   \n",
       "173485197  ffffeba9dbf2bab1     chrY:19499847-19500746      1   \n",
       "173485198  ffffeba9dbf2bab1     chrY:19566990-19567754      1   \n",
       "173485199  ffffeba9dbf2bab1     chrY:20735890-20736626      3   \n",
       "173485200  ffffeba9dbf2bab1     chrY:21260054-21260842      2   \n",
       "173485201  ffffeba9dbf2bab1       chrY:7273431-7274301      2   \n",
       "\n",
       "           normalized_count chromosome      start        end  \n",
       "0                  2.664076      chr10  100346599  100347467  \n",
       "1                  4.966185      chr10  101214564  101215355  \n",
       "2                  1.627825      chr10  101353529  101354425  \n",
       "3                  1.333610      chr10  101830364  101831303  \n",
       "4                  1.588739      chr10  102064862  102065750  \n",
       "...                     ...        ...        ...        ...  \n",
       "173485197          3.342186       chrY   19499847   19500746  \n",
       "173485198          1.154666       chrY   19566990   19567754  \n",
       "173485199          7.049924       chrY   20735890   20736626  \n",
       "173485200          3.967924       chrY   21260054   21260842  \n",
       "173485201          1.250685       chrY    7273431    7274301  \n",
       "\n",
       "[173485202 rows x 7 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiome_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "multiome_obs_meta = pd.read_csv(\"../../data/multiome_obs_meta.csv\")\n",
    "B_cells_obs_ids = multiome_obs_meta[multiome_obs_meta[\"cell_type\"] == \"B cells\"][\"obs_id\"].to_list()\n",
    "T_cells_CD4_obs_ids = multiome_obs_meta[multiome_obs_meta[\"cell_type\"] == \"T cells CD4+\"][\"obs_id\"].to_list()\n",
    "T_cells_CD8_obs_ids = multiome_obs_meta[multiome_obs_meta[\"cell_type\"] == \"T cells CD8+\"][\"obs_id\"].to_list()\n",
    "Myeloid_obs_ids = multiome_obs_meta[multiome_obs_meta[\"cell_type\"] == \"Myeloid cells\"][\"obs_id\"].to_list()\n",
    "NK_obs_ids = multiome_obs_meta[multiome_obs_meta[\"cell_type\"] == \"NK cells\"][\"obs_id\"].to_list()\n",
    "T_regulatory_obs_ids = multiome_obs_meta[multiome_obs_meta[\"cell_type\"] == \"T regulatory cells\"][\"obs_id\"].to_list()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "B_cells_atac = multiome_df[multiome_df[\"obs_id\"].isin(B_cells_obs_ids)]\n",
    "T_cells_CD4_atac = multiome_df[multiome_df[\"obs_id\"].isin(T_cells_CD4_obs_ids)]\n",
    "T_cells_CD8_atac = multiome_df[multiome_df[\"obs_id\"].isin(T_cells_CD8_obs_ids)]\n",
    "Myeloid_atac = multiome_df[multiome_df[\"obs_id\"].isin(Myeloid_obs_ids)]\n",
    "NK_atac = multiome_df[multiome_df[\"obs_id\"].isin(NK_obs_ids)]\n",
    "T_regulatory_atac = multiome_df[multiome_df[\"obs_id\"].isin(T_regulatory_obs_ids)] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "correlations = pd.read_table(\"../../data/allGeneCorrelations100000.p05_v3.txt\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_chr</th>\n",
       "      <th>dhs_start</th>\n",
       "      <th>dhs_end</th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>metaprobeset_id</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>cor</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99861860</td>\n",
       "      <td>99862010</td>\n",
       "      <td>2861605</td>\n",
       "      <td>chrX</td>\n",
       "      <td>99885757</td>\n",
       "      <td>99891766</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>-0.544568</td>\n",
       "      <td>0.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99880960</td>\n",
       "      <td>99881110</td>\n",
       "      <td>2861619</td>\n",
       "      <td>chrX</td>\n",
       "      <td>99885757</td>\n",
       "      <td>99891766</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>-0.402630</td>\n",
       "      <td>0.020</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99942220</td>\n",
       "      <td>99942370</td>\n",
       "      <td>2861690</td>\n",
       "      <td>chrX</td>\n",
       "      <td>99885757</td>\n",
       "      <td>99891766</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>-0.385988</td>\n",
       "      <td>0.023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99942700</td>\n",
       "      <td>99942850</td>\n",
       "      <td>2861691</td>\n",
       "      <td>chrX</td>\n",
       "      <td>99885757</td>\n",
       "      <td>99891766</td>\n",
       "      <td>TSPAN6</td>\n",
       "      <td>1</td>\n",
       "      <td>ENSG00000000003</td>\n",
       "      <td>-0.446240</td>\n",
       "      <td>0.016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>chrX</td>\n",
       "      <td>99870220</td>\n",
       "      <td>99870370</td>\n",
       "      <td>2861610</td>\n",
       "      <td>chrX</td>\n",
       "      <td>99839836</td>\n",
       "      <td>99854854</td>\n",
       "      <td>TNMD</td>\n",
       "      <td>2</td>\n",
       "      <td>ENSG00000000005</td>\n",
       "      <td>-0.065758</td>\n",
       "      <td>0.017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793518</th>\n",
       "      <td>chr15</td>\n",
       "      <td>83781325</td>\n",
       "      <td>83781475</td>\n",
       "      <td>931308</td>\n",
       "      <td>chr15</td>\n",
       "      <td>83691029</td>\n",
       "      <td>83691067</td>\n",
       "      <td>RP11-90B9.2</td>\n",
       "      <td>52291</td>\n",
       "      <td>ENSG00000259767</td>\n",
       "      <td>-0.169041</td>\n",
       "      <td>0.024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793519</th>\n",
       "      <td>chr15</td>\n",
       "      <td>83781480</td>\n",
       "      <td>83781630</td>\n",
       "      <td>931309</td>\n",
       "      <td>chr15</td>\n",
       "      <td>83691029</td>\n",
       "      <td>83691067</td>\n",
       "      <td>RP11-90B9.2</td>\n",
       "      <td>52291</td>\n",
       "      <td>ENSG00000259767</td>\n",
       "      <td>-0.245031</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793526</th>\n",
       "      <td>chr16</td>\n",
       "      <td>72821420</td>\n",
       "      <td>72821570</td>\n",
       "      <td>1025241</td>\n",
       "      <td>chr16</td>\n",
       "      <td>72779322</td>\n",
       "      <td>72781064</td>\n",
       "      <td>RP5-991G20.1</td>\n",
       "      <td>52292</td>\n",
       "      <td>ENSG00000259768</td>\n",
       "      <td>-0.253807</td>\n",
       "      <td>0.021</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793547</th>\n",
       "      <td>chr15</td>\n",
       "      <td>31528800</td>\n",
       "      <td>31528950</td>\n",
       "      <td>868131</td>\n",
       "      <td>chr15</td>\n",
       "      <td>31514639</td>\n",
       "      <td>31522962</td>\n",
       "      <td>RP11-16E12.2</td>\n",
       "      <td>52296</td>\n",
       "      <td>ENSG00000259772</td>\n",
       "      <td>-0.320208</td>\n",
       "      <td>0.005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>793550</th>\n",
       "      <td>chr15</td>\n",
       "      <td>31546860</td>\n",
       "      <td>31547010</td>\n",
       "      <td>868160</td>\n",
       "      <td>chr15</td>\n",
       "      <td>31514639</td>\n",
       "      <td>31522962</td>\n",
       "      <td>RP11-16E12.2</td>\n",
       "      <td>52296</td>\n",
       "      <td>ENSG00000259772</td>\n",
       "      <td>-0.236022</td>\n",
       "      <td>0.018</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176352 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       dhs_chr  dhs_start   dhs_end   dhs_id gene_chr  gene_start  gene_end  \\\n",
       "0         chrX   99861860  99862010  2861605     chrX    99885757  99891766   \n",
       "2         chrX   99880960  99881110  2861619     chrX    99885757  99891766   \n",
       "17        chrX   99942220  99942370  2861690     chrX    99885757  99891766   \n",
       "18        chrX   99942700  99942850  2861691     chrX    99885757  99891766   \n",
       "24        chrX   99870220  99870370  2861610     chrX    99839836  99854854   \n",
       "...        ...        ...       ...      ...      ...         ...       ...   \n",
       "793518   chr15   83781325  83781475   931308    chr15    83691029  83691067   \n",
       "793519   chr15   83781480  83781630   931309    chr15    83691029  83691067   \n",
       "793526   chr16   72821420  72821570  1025241    chr16    72779322  72781064   \n",
       "793547   chr15   31528800  31528950   868131    chr15    31514639  31522962   \n",
       "793550   chr15   31546860  31547010   868160    chr15    31514639  31522962   \n",
       "\n",
       "           gene_name  metaprobeset_id        ensemblID       cor   pval  \n",
       "0             TSPAN6                1  ENSG00000000003 -0.544568  0.006  \n",
       "2             TSPAN6                1  ENSG00000000003 -0.402630  0.020  \n",
       "17            TSPAN6                1  ENSG00000000003 -0.385988  0.023  \n",
       "18            TSPAN6                1  ENSG00000000003 -0.446240  0.016  \n",
       "24              TNMD                2  ENSG00000000005 -0.065758  0.017  \n",
       "...              ...              ...              ...       ...    ...  \n",
       "793518   RP11-90B9.2            52291  ENSG00000259767 -0.169041  0.024  \n",
       "793519   RP11-90B9.2            52291  ENSG00000259767 -0.245031  0.004  \n",
       "793526  RP5-991G20.1            52292  ENSG00000259768 -0.253807  0.021  \n",
       "793547  RP11-16E12.2            52296  ENSG00000259772 -0.320208  0.005  \n",
       "793550  RP11-16E12.2            52296  ENSG00000259772 -0.236022  0.018  \n",
       "\n",
       "[176352 rows x 12 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_correlations = correlations[correlations[\"pval\"]<=0.05]\n",
    "sig_correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chrX'"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_correlations.iloc[0,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ensemble_ids = list(set(list(sig_correlations[\"ensemblID\"])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_chr</th>\n",
       "      <th>dhs_start</th>\n",
       "      <th>dhs_end</th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>metaprobeset_id</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>cor</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>634736</th>\n",
       "      <td>chr7</td>\n",
       "      <td>10080</td>\n",
       "      <td>10230</td>\n",
       "      <td>2398536</td>\n",
       "      <td>chr7</td>\n",
       "      <td>71699</td>\n",
       "      <td>71819</td>\n",
       "      <td>AC093627.7</td>\n",
       "      <td>33334</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>-0.343532</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481648</th>\n",
       "      <td>chr17</td>\n",
       "      <td>14460</td>\n",
       "      <td>14610</td>\n",
       "      <td>1051511</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31380</td>\n",
       "      <td>31404</td>\n",
       "      <td>DOC2B</td>\n",
       "      <td>16908</td>\n",
       "      <td>ENSG00000187939</td>\n",
       "      <td>-0.244659</td>\n",
       "      <td>0.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>683753</th>\n",
       "      <td>chr1</td>\n",
       "      <td>16140</td>\n",
       "      <td>16290</td>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>63656</td>\n",
       "      <td>63697</td>\n",
       "      <td>OR4G11P</td>\n",
       "      <td>39505</td>\n",
       "      <td>ENSG00000240361</td>\n",
       "      <td>-0.223318</td>\n",
       "      <td>0.011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>695472</th>\n",
       "      <td>chr7</td>\n",
       "      <td>37380</td>\n",
       "      <td>37530</td>\n",
       "      <td>2398545</td>\n",
       "      <td>chr7</td>\n",
       "      <td>135911</td>\n",
       "      <td>149427</td>\n",
       "      <td>AC093627.9</td>\n",
       "      <td>41017</td>\n",
       "      <td>ENSG00000242474</td>\n",
       "      <td>-0.254091</td>\n",
       "      <td>0.011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>634737</th>\n",
       "      <td>chr7</td>\n",
       "      <td>42445</td>\n",
       "      <td>42595</td>\n",
       "      <td>2398553</td>\n",
       "      <td>chr7</td>\n",
       "      <td>71699</td>\n",
       "      <td>71819</td>\n",
       "      <td>AC093627.7</td>\n",
       "      <td>33334</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>-0.307110</td>\n",
       "      <td>0.014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>516698</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249201360</td>\n",
       "      <td>249201510</td>\n",
       "      <td>262476</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249206914</td>\n",
       "      <td>249206945</td>\n",
       "      <td>U6</td>\n",
       "      <td>19119</td>\n",
       "      <td>ENSG00000200495</td>\n",
       "      <td>-0.068287</td>\n",
       "      <td>0.017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>416425</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249202585</td>\n",
       "      <td>249202735</td>\n",
       "      <td>262479</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249106105</td>\n",
       "      <td>249111073</td>\n",
       "      <td>SH3BP5L</td>\n",
       "      <td>14012</td>\n",
       "      <td>ENSG00000175137</td>\n",
       "      <td>-0.185339</td>\n",
       "      <td>0.014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392632</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249202785</td>\n",
       "      <td>249202935</td>\n",
       "      <td>262480</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249144235</td>\n",
       "      <td>249153279</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>13161</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>-0.258991</td>\n",
       "      <td>0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392633</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249212885</td>\n",
       "      <td>249213035</td>\n",
       "      <td>262509</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249144235</td>\n",
       "      <td>249153279</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>13161</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>-0.258949</td>\n",
       "      <td>0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>392634</th>\n",
       "      <td>chr1</td>\n",
       "      <td>249218165</td>\n",
       "      <td>249218315</td>\n",
       "      <td>262513</td>\n",
       "      <td>chr1</td>\n",
       "      <td>249144235</td>\n",
       "      <td>249153279</td>\n",
       "      <td>ZNF692</td>\n",
       "      <td>13161</td>\n",
       "      <td>ENSG00000171163</td>\n",
       "      <td>-0.272388</td>\n",
       "      <td>0.017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176352 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       dhs_chr  dhs_start    dhs_end   dhs_id gene_chr  gene_start   gene_end  \\\n",
       "634736    chr7      10080      10230  2398536     chr7       71699      71819   \n",
       "481648   chr17      14460      14610  1051511    chr17       31380      31404   \n",
       "683753    chr1      16140      16290        3     chr1       63656      63697   \n",
       "695472    chr7      37380      37530  2398545     chr7      135911     149427   \n",
       "634737    chr7      42445      42595  2398553     chr7       71699      71819   \n",
       "...        ...        ...        ...      ...      ...         ...        ...   \n",
       "516698    chr1  249201360  249201510   262476     chr1   249206914  249206945   \n",
       "416425    chr1  249202585  249202735   262479     chr1   249106105  249111073   \n",
       "392632    chr1  249202785  249202935   262480     chr1   249144235  249153279   \n",
       "392633    chr1  249212885  249213035   262509     chr1   249144235  249153279   \n",
       "392634    chr1  249218165  249218315   262513     chr1   249144235  249153279   \n",
       "\n",
       "         gene_name  metaprobeset_id        ensemblID       cor   pval  \n",
       "634736  AC093627.7            33334  ENSG00000232325 -0.343532  0.003  \n",
       "481648       DOC2B            16908  ENSG00000187939 -0.244659  0.010  \n",
       "683753     OR4G11P            39505  ENSG00000240361 -0.223318  0.011  \n",
       "695472  AC093627.9            41017  ENSG00000242474 -0.254091  0.011  \n",
       "634737  AC093627.7            33334  ENSG00000232325 -0.307110  0.014  \n",
       "...            ...              ...              ...       ...    ...  \n",
       "516698          U6            19119  ENSG00000200495 -0.068287  0.017  \n",
       "416425     SH3BP5L            14012  ENSG00000175137 -0.185339  0.014  \n",
       "392632      ZNF692            13161  ENSG00000171163 -0.258991  0.019  \n",
       "392633      ZNF692            13161  ENSG00000171163 -0.258949  0.019  \n",
       "392634      ZNF692            13161  ENSG00000171163 -0.272388  0.017  \n",
       "\n",
       "[176352 rows x 12 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_correlations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\ghanb\\AppData\\Local\\Temp\\ipykernel_25788\\2579570082.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  sig_correlations.sort_values(by=\"dhs_start\", inplace=True)\n"
     ]
    }
   ],
   "source": [
    "sig_correlations.sort_values(by=\"dhs_start\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr7'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_correlations[\"dhs_chr\"].unique()[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\ghanb\\AppData\\Local\\Temp\\ipykernel_25788\\159472389.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  B_cells_atac.sort_values(by=\"start\", inplace=True)\n"
     ]
    }
   ],
   "source": [
    "B_cells_atac.sort_values(by=\"start\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "start      end      \n",
       "2090       2982         2.288582\n",
       "3945       4846         3.229936\n",
       "4505       5401         3.923541\n",
       "7910       8736         3.159859\n",
       "9685       10586        3.067677\n",
       "                          ...   \n",
       "248866924  248867879    2.411894\n",
       "248873533  248874443    0.886628\n",
       "248906099  248906933    0.577764\n",
       "248924630  248925480    1.573709\n",
       "248945354  248946091    2.421512\n",
       "Name: normalized_count, Length: 135154, dtype: float32"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "B_cells_atac.groupby(by=[\"start\", \"end\"])[\"normalized_count\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gene_atac(atac, cor, ensemble_ids):\n",
    "    row = 0\n",
    "    gene_atac = pd.DataFrame(columns=[\"obs_id\", \"location\", \"chromosome\", \"ensembleID\", \"intersection\", \"cor_start\", \"cor_end\", \"score\"])\n",
    "    for id in tqdm(ensemble_ids):\n",
    "        gene_ranges = cor[cor[\"ensemblID\"]==id]\n",
    "        chromosome = str(list(gene_ranges[\"dhs_chr\"])[0])\n",
    "        atac = atac[atac[\"chromosome\"]==chromosome]\n",
    "        for i in range(len(gene_ranges)):\n",
    "            seq_range = range(int(gene_ranges[\"dhs_start\"].iloc[i]), int(gene_ranges[\"dhs_end\"].iloc[i])+1)\n",
    "\n",
    "            for j in range(len(atac)):\n",
    "                location_range = range(int(atac[\"start\"].iloc[j]), int(atac[\"end\"].iloc[j])+1)\n",
    "                ranges_intersection = len(set(seq_range).intersection(location_range))\n",
    "                if ranges_intersection>0:\n",
    "                    gene_atac.iloc[row, 0] = atac[\"obs_id\"].iloc[j]\n",
    "                    gene_atac.iloc[row, 1] = atac[\"location\"].iloc[j]\n",
    "                    gene_atac.iloc[row, 2] = atac[\"chromosome\"].iloc[j]\n",
    "                    gene_atac.iloc[row, 3] = id\n",
    "                    gene_atac.iloc[row, 4] = ranges_intersection/len(seq_range)\n",
    "                    gene_atac.iloc[row, 5] = gene_ranges[\"dhs_start\"].iloc[i]\n",
    "                    gene_atac.iloc[row, 6] = gene_ranges[\"dhs_end\"].iloc[i]\n",
    "                    gene_atac.iloc[row, 7] = atac[\"normalized_count\"].iloc[j]\n",
    "\n",
    "                    row+=1\n",
    "                    \n",
    "    return gene_atac        \n",
    "                    \n",
    "                \n",
    "            \n",
    "            \n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gene_atac2(atac, cor):\n",
    "    gene_atac = pd.DataFrame(np.zeros((len(cor), 7)))\n",
    "    gene_atac.columns = [\"obs_id\", \"chromosome\", \"ensembleID\", \"intersection\", \"cor_start\", \"cor_end\", \"score\"]\n",
    "    all_cor = cor.copy()\n",
    "    all_atac = atac.copy()\n",
    "    gene_atac_row = 0\n",
    "\n",
    "    for chromosome in tqdm(all_cor[\"dhs_chr\"].unique()):\n",
    "        atac = all_atac[all_atac[\"chromosome\"]==chromosome]\n",
    "        cor = all_cor[all_cor[\"dhs_chr\"]==chromosome]\n",
    "\n",
    "        pointer = 0\n",
    "        direction = -1\n",
    "        \n",
    "        for i, atac_row in atac.iterrows():\n",
    "            direction*=-1\n",
    "            atac_start = atac_row[\"start\"]\n",
    "            atac_end = atac_row[\"end\"]   \n",
    "            flag = 0 \n",
    "              \n",
    "            while pointer<len(cor) and flag !=2 and gene_atac_row<=len(all_cor):\n",
    "                #if gene_atac_row==len(all_cor)+1:\n",
    "                    #return gene_atac\n",
    "                if atac_start < cor[\"dhs_end\"].iloc[pointer] and atac_end > cor[\"dhs_start\"].iloc[pointer]:\n",
    "                    flag = 1\n",
    "                    seq_range = range(int(cor[\"dhs_start\"].iloc[pointer]), int(cor[\"dhs_end\"].iloc[pointer])+1)\n",
    "                    location_range = range(atac_start, atac_end+1)\n",
    "                    ranges_intersection = len(set(seq_range).intersection(location_range))\n",
    "\n",
    "                    gene_atac.iloc[gene_atac_row, 0] = atac_row[\"obs_id\"]\n",
    "                    gene_atac.iloc[gene_atac_row, 1] = chromosome\n",
    "                    gene_atac.iloc[gene_atac_row, 2] = cor[\"ensemblID\"].iloc[pointer]\n",
    "                    gene_atac.iloc[gene_atac_row, 3] = ranges_intersection/len(seq_range)\n",
    "                    gene_atac.iloc[gene_atac_row, 4] = cor[\"dhs_start\"].iloc[pointer]\n",
    "                    gene_atac.iloc[gene_atac_row, 5] = cor[\"dhs_end\"].iloc[pointer]\n",
    "                    gene_atac.iloc[gene_atac_row, 6] = atac_row[\"normalized_count\"]\n",
    "\n",
    "                    pointer+=1\n",
    "                    gene_atac_row+=1\n",
    "                    print(pointer)\n",
    "                else:\n",
    "                    if flag == 0:\n",
    "                        \n",
    "                        if direction==-1:\n",
    "                            pointer+=1\n",
    "                        else:\n",
    "                            pointer-=1\n",
    "                        \n",
    "                    else:\n",
    "                        flag = 2\n",
    "    return gene_atac\n",
    "    \n",
    "    \n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_gene_atac3(atac, cor):\n",
    "    gene_atac = pd.DataFrame(columns=[\"obs_id\", \"chromosome\", \"ensembleID\", \"intersection\", \"score\"])\n",
    "    # gene_atac.columns = [\"obs_id\", \"chromosome\", \"ensembleID\", \"intersection\", \"cor_start\", \"cor_end\", \"score\"]\n",
    "    all_cor = cor.copy()\n",
    "    all_atac = atac.copy()\n",
    "\n",
    "    for cn, chromosome in enumerate(all_cor[\"dhs_chr\"].unique()[9:11]):\n",
    "        atac = all_atac[all_atac[\"chromosome\"]==chromosome]\n",
    "        cor = all_cor[all_cor[\"dhs_chr\"]==chromosome]\n",
    "        print(f\"chromosome no: {cn} | name: {chromosome}\")\n",
    "        print(f\"atac len for chromosome: {len(atac)}\")\n",
    "        print(f\"cor len for chromosome: {len(cor)}\")\n",
    "\n",
    "        p_atac, p_cor = 0, 0\n",
    "\n",
    "        ### ALIGN\n",
    "        ###\n",
    "\n",
    "        if cor[\"dhs_start\"].iloc[p_cor] > atac[\"end\"].iloc[p_atac]:\n",
    "            # while not (atac[\"start\"].iloc[p_atac] < cor[\"dhs_end\"].iloc[p_cor] and \\\n",
    "            #            atac[\"end\"].iloc[p_atac] > cor[\"dhs_start\"].iloc[p_cor]):\n",
    "            while cor[\"dhs_start\"].iloc[p_cor] > atac[\"end\"].iloc[p_atac]:\n",
    "                p_atac += 1\n",
    "        elif cor[\"dhs_end\"].iloc[p_cor] < atac[\"start\"].iloc[p_atac]:\n",
    "            # while not (atac[\"start\"].iloc[p_atac] < cor[\"dhs_end\"].iloc[p_cor] and \\\n",
    "            #            atac[\"end\"].iloc[p_atac] > cor[\"dhs_start\"].iloc[p_cor]):\n",
    "            while cor[\"dhs_end\"].iloc[p_cor] < atac[\"start\"].iloc[p_atac]:\n",
    "                p_cor += 1\n",
    "\n",
    "        while p_cor < len(cor) and p_atac < len(atac): # shart e kharej shodan\n",
    "            \n",
    "            \n",
    "            if p_cor%1000==0:\n",
    "                print(p_cor, end=\", \")\n",
    "            while atac[\"start\"].iloc[p_atac] < cor[\"dhs_end\"].iloc[p_cor] and \\\n",
    "                atac[\"end\"].iloc[p_atac] > cor[\"dhs_start\"].iloc[p_cor]:\n",
    "\n",
    "                \n",
    "                obs_id = atac[\"obs_id\"].iloc[p_atac]\n",
    "                # chromosome\n",
    "                ensemble_id = cor[\"ensemblID\"].iloc[p_cor]\n",
    "                dhs_start = cor[\"dhs_start\"].iloc[p_cor]\n",
    "                dhs_end = cor[\"dhs_end\"].iloc[p_cor]\n",
    "                atac_start = atac[\"start\"].iloc[p_atac]\n",
    "                atac_end = atac[\"end\"].iloc[p_atac]\n",
    "                intersection = (min(atac_end, dhs_end)-(max(atac_start, dhs_start)))/(dhs_end-dhs_start)\n",
    "\n",
    "                normalized_count = atac[\"normalized_count\"].iloc[p_atac]\n",
    "                series = pd.DataFrame(\n",
    "                    {\n",
    "                        \"obs_id\": [obs_id],\n",
    "                        \"chromosome\": [chromosome],\n",
    "                        \"ensembleID\": [ensemble_id],\n",
    "                        \"intersection\": [intersection],\n",
    "                        \"score\": [normalized_count]\n",
    "                    }\n",
    "                )\n",
    "                gene_atac = pd.concat([gene_atac, series], axis=0)\n",
    "                \n",
    "\n",
    "                p_cor += 1\n",
    "                if p_cor==len(cor):\n",
    "                    p_cor-=1\n",
    "                    break\n",
    "            p_atac += 1\n",
    "            if p_atac==len(atac):\n",
    "                p_atac-=1\n",
    "                break\n",
    "            while atac[\"start\"].iloc[p_atac]==atac[\"start\"].iloc[p_atac-1] and \\\n",
    "                atac[\"end\"].iloc[p_atac]==atac[\"end\"].iloc[p_atac-1]:\n",
    "                p_atac+=1\n",
    "                if p_atac==len(atac):\n",
    "                    p_atac-=1\n",
    "                    break\n",
    "\n",
    "            if p_atac == len(atac):\n",
    "                break\n",
    "\n",
    "            if cor[\"dhs_start\"].iloc[p_cor] > atac[\"end\"].iloc[p_atac]:\n",
    "                while cor[\"dhs_start\"].iloc[p_cor] > atac[\"end\"].iloc[p_atac]:\n",
    "                    # print(\"hiii\", p_cor)\n",
    "                    p_cor -= 1\n",
    "                while atac[\"start\"].iloc[p_atac] < cor[\"dhs_end\"].iloc[p_cor] and \\\n",
    "                    atac[\"end\"].iloc[p_atac] > cor[\"dhs_start\"].iloc[p_cor]:\n",
    "                    p_cor -= 1\n",
    "                p_cor += 1\n",
    "                if p_cor==len(cor):\n",
    "                    p_cor-=1\n",
    "                    break\n",
    "                continue\n",
    "            if cor[\"dhs_end\"].iloc[p_cor] < atac[\"start\"].iloc[p_atac]:\n",
    "                while cor[\"dhs_end\"].iloc[p_cor] < atac[\"start\"].iloc[p_atac]:\n",
    "                    p_cor += 1\n",
    "                    if p_cor == len(cor):\n",
    "                        p_cor -= 1\n",
    "                        break\n",
    "                    \n",
    "                continue\n",
    "            while atac[\"start\"].iloc[p_atac] < cor[\"dhs_end\"].iloc[p_cor] and \\\n",
    "                atac[\"end\"].iloc[p_atac] > cor[\"dhs_start\"].iloc[p_cor]:\n",
    "                p_cor -= 1\n",
    "            p_cor += 1\n",
    "            if p_cor == len(cor):\n",
    "                p_cor -= 1\n",
    "                break\n",
    "            continue\n",
    "    \n",
    "    return gene_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "chromosome no: 0 | name: chr10\n",
      "atac len for chromosome: 1224389\n",
      "cor len for chromosome: 6753\n",
      "0, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, 4000, chromosome no: 1 | name: chr6\n",
      "atac len for chromosome: 1704688\n",
      "cor len for chromosome: 9489\n",
      "0, 4000, 8000, 8000, 8000, "
     ]
    }
   ],
   "source": [
    "B_cells_gene_atac = create_gene_atac3(B_cells_atac, sig_correlations)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>ensembleID</th>\n",
       "      <th>intersection</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>05fadfb8107d4e0c</td>\n",
       "      <td>chr10</td>\n",
       "      <td>ENSG00000151240</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>2.097998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>7af4f888f5661a62</td>\n",
       "      <td>chr10</td>\n",
       "      <td>ENSG00000239822</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.443676</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1a69bd9e461e00b7</td>\n",
       "      <td>chr10</td>\n",
       "      <td>ENSG00000047056</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.932475</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>52e77bb794556848</td>\n",
       "      <td>chr10</td>\n",
       "      <td>ENSG00000237399</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.760328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21462ae0a553a6ae</td>\n",
       "      <td>chr10</td>\n",
       "      <td>ENSG00000236892</td>\n",
       "      <td>0.820000</td>\n",
       "      <td>3.683643</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2bb210b00604ce9e</td>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000184786</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.880386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2bb210b00604ce9e</td>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000130024</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.880386</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>c0c2d3413321e293</td>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000185127</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.647786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2b8b1ac60e0f82e0</td>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000236173</td>\n",
       "      <td>0.820000</td>\n",
       "      <td>1.507023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>89d6ff4dfa20815b</td>\n",
       "      <td>chr6</td>\n",
       "      <td>ENSG00000231690</td>\n",
       "      <td>0.146667</td>\n",
       "      <td>0.925316</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1662 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              obs_id chromosome       ensembleID  intersection     score\n",
       "0   05fadfb8107d4e0c      chr10  ENSG00000151240      1.000000  2.097998\n",
       "0   7af4f888f5661a62      chr10  ENSG00000239822      1.000000  1.443676\n",
       "0   1a69bd9e461e00b7      chr10  ENSG00000047056      1.000000  0.932475\n",
       "0   52e77bb794556848      chr10  ENSG00000237399      1.000000  0.760328\n",
       "0   21462ae0a553a6ae      chr10  ENSG00000236892      0.820000  3.683643\n",
       "..               ...        ...              ...           ...       ...\n",
       "0   2bb210b00604ce9e       chr6  ENSG00000184786      1.000000  0.880386\n",
       "0   2bb210b00604ce9e       chr6  ENSG00000130024      1.000000  0.880386\n",
       "0   c0c2d3413321e293       chr6  ENSG00000185127      1.000000  1.647786\n",
       "0   2b8b1ac60e0f82e0       chr6  ENSG00000236173      0.820000  1.507023\n",
       "0   89d6ff4dfa20815b       chr6  ENSG00000231690      0.146667  0.925316\n",
       "\n",
       "[1662 rows x 5 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "B_cells_gene_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>location</th>\n",
       "      <th>count</th>\n",
       "      <th>normalized_count</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>134540387</th>\n",
       "      <td>c50c5ab93f69ce46</td>\n",
       "      <td>chr17:112841-113741</td>\n",
       "      <td>2</td>\n",
       "      <td>1.819549</td>\n",
       "      <td>chr17</td>\n",
       "      <td>112841</td>\n",
       "      <td>113741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138524130</th>\n",
       "      <td>cb8174035d4e5865</td>\n",
       "      <td>chr17:112841-113741</td>\n",
       "      <td>2</td>\n",
       "      <td>1.480782</td>\n",
       "      <td>chr17</td>\n",
       "      <td>112841</td>\n",
       "      <td>113741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>127592517</th>\n",
       "      <td>ba7b8dc1116cf085</td>\n",
       "      <td>chr17:112841-113741</td>\n",
       "      <td>2</td>\n",
       "      <td>1.734908</td>\n",
       "      <td>chr17</td>\n",
       "      <td>112841</td>\n",
       "      <td>113741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>151103920</th>\n",
       "      <td>de976dabfb43f494</td>\n",
       "      <td>chr17:112841-113741</td>\n",
       "      <td>2</td>\n",
       "      <td>1.121243</td>\n",
       "      <td>chr17</td>\n",
       "      <td>112841</td>\n",
       "      <td>113741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9292552</th>\n",
       "      <td>0d6386d9fe2cd66d</td>\n",
       "      <td>chr17:112841-113741</td>\n",
       "      <td>2</td>\n",
       "      <td>1.573857</td>\n",
       "      <td>chr17</td>\n",
       "      <td>112841</td>\n",
       "      <td>113741</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>153582610</th>\n",
       "      <td>e23ab6cb65282945</td>\n",
       "      <td>chr17:83246798-83247700</td>\n",
       "      <td>2</td>\n",
       "      <td>3.337099</td>\n",
       "      <td>chr17</td>\n",
       "      <td>83246798</td>\n",
       "      <td>83247700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132216518</th>\n",
       "      <td>c18ef8ec5c239e07</td>\n",
       "      <td>chr17:83246798-83247700</td>\n",
       "      <td>2</td>\n",
       "      <td>1.937325</td>\n",
       "      <td>chr17</td>\n",
       "      <td>83246798</td>\n",
       "      <td>83247700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5941064</th>\n",
       "      <td>08b42f34074d30cd</td>\n",
       "      <td>chr17:83246798-83247700</td>\n",
       "      <td>2</td>\n",
       "      <td>2.060075</td>\n",
       "      <td>chr17</td>\n",
       "      <td>83246798</td>\n",
       "      <td>83247700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38964356</th>\n",
       "      <td>387dde2de6c538b8</td>\n",
       "      <td>chr17:83246798-83247700</td>\n",
       "      <td>2</td>\n",
       "      <td>2.254571</td>\n",
       "      <td>chr17</td>\n",
       "      <td>83246798</td>\n",
       "      <td>83247700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>138373877</th>\n",
       "      <td>cb52850fcab2e7d7</td>\n",
       "      <td>chr17:83246798-83247700</td>\n",
       "      <td>2</td>\n",
       "      <td>1.573529</td>\n",
       "      <td>chr17</td>\n",
       "      <td>83246798</td>\n",
       "      <td>83247700</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1773105 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     obs_id                 location  count  normalized_count  \\\n",
       "134540387  c50c5ab93f69ce46      chr17:112841-113741      2          1.819549   \n",
       "138524130  cb8174035d4e5865      chr17:112841-113741      2          1.480782   \n",
       "127592517  ba7b8dc1116cf085      chr17:112841-113741      2          1.734908   \n",
       "151103920  de976dabfb43f494      chr17:112841-113741      2          1.121243   \n",
       "9292552    0d6386d9fe2cd66d      chr17:112841-113741      2          1.573857   \n",
       "...                     ...                      ...    ...               ...   \n",
       "153582610  e23ab6cb65282945  chr17:83246798-83247700      2          3.337099   \n",
       "132216518  c18ef8ec5c239e07  chr17:83246798-83247700      2          1.937325   \n",
       "5941064    08b42f34074d30cd  chr17:83246798-83247700      2          2.060075   \n",
       "38964356   387dde2de6c538b8  chr17:83246798-83247700      2          2.254571   \n",
       "138373877  cb52850fcab2e7d7  chr17:83246798-83247700      2          1.573529   \n",
       "\n",
       "          chromosome     start       end  \n",
       "134540387      chr17    112841    113741  \n",
       "138524130      chr17    112841    113741  \n",
       "127592517      chr17    112841    113741  \n",
       "151103920      chr17    112841    113741  \n",
       "9292552        chr17    112841    113741  \n",
       "...              ...       ...       ...  \n",
       "153582610      chr17  83246798  83247700  \n",
       "132216518      chr17  83246798  83247700  \n",
       "5941064        chr17  83246798  83247700  \n",
       "38964356       chr17  83246798  83247700  \n",
       "138373877      chr17  83246798  83247700  \n",
       "\n",
       "[1773105 rows x 7 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "B_cells_atac[B_cells_atac[\"chromosome\"]==\"chr17\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dhs_chr</th>\n",
       "      <th>dhs_start</th>\n",
       "      <th>dhs_end</th>\n",
       "      <th>dhs_id</th>\n",
       "      <th>gene_chr</th>\n",
       "      <th>gene_start</th>\n",
       "      <th>gene_end</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>metaprobeset_id</th>\n",
       "      <th>ensemblID</th>\n",
       "      <th>cor</th>\n",
       "      <th>pval</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>481648</th>\n",
       "      <td>chr17</td>\n",
       "      <td>14460</td>\n",
       "      <td>14610</td>\n",
       "      <td>1051511</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31380</td>\n",
       "      <td>31404</td>\n",
       "      <td>DOC2B</td>\n",
       "      <td>16908</td>\n",
       "      <td>ENSG00000187939</td>\n",
       "      <td>-0.244659</td>\n",
       "      <td>0.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481651</th>\n",
       "      <td>chr17</td>\n",
       "      <td>49180</td>\n",
       "      <td>49330</td>\n",
       "      <td>1051584</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31380</td>\n",
       "      <td>31404</td>\n",
       "      <td>DOC2B</td>\n",
       "      <td>16908</td>\n",
       "      <td>ENSG00000187939</td>\n",
       "      <td>-0.242436</td>\n",
       "      <td>0.010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481652</th>\n",
       "      <td>chr17</td>\n",
       "      <td>49800</td>\n",
       "      <td>49950</td>\n",
       "      <td>1051585</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31380</td>\n",
       "      <td>31404</td>\n",
       "      <td>DOC2B</td>\n",
       "      <td>16908</td>\n",
       "      <td>ENSG00000187939</td>\n",
       "      <td>-0.308090</td>\n",
       "      <td>0.003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481653</th>\n",
       "      <td>chr17</td>\n",
       "      <td>49980</td>\n",
       "      <td>50130</td>\n",
       "      <td>1051586</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31380</td>\n",
       "      <td>31404</td>\n",
       "      <td>DOC2B</td>\n",
       "      <td>16908</td>\n",
       "      <td>ENSG00000187939</td>\n",
       "      <td>-0.260791</td>\n",
       "      <td>0.006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481654</th>\n",
       "      <td>chr17</td>\n",
       "      <td>50160</td>\n",
       "      <td>50310</td>\n",
       "      <td>1051587</td>\n",
       "      <td>chr17</td>\n",
       "      <td>31380</td>\n",
       "      <td>31404</td>\n",
       "      <td>DOC2B</td>\n",
       "      <td>16908</td>\n",
       "      <td>ENSG00000187939</td>\n",
       "      <td>-0.287450</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>424856</th>\n",
       "      <td>chr17</td>\n",
       "      <td>81128545</td>\n",
       "      <td>81128695</td>\n",
       "      <td>1167186</td>\n",
       "      <td>chr17</td>\n",
       "      <td>81037578</td>\n",
       "      <td>81043172</td>\n",
       "      <td>METRNL</td>\n",
       "      <td>14362</td>\n",
       "      <td>ENSG00000176845</td>\n",
       "      <td>-0.329763</td>\n",
       "      <td>0.022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638640</th>\n",
       "      <td>chr17</td>\n",
       "      <td>81137460</td>\n",
       "      <td>81137610</td>\n",
       "      <td>1167210</td>\n",
       "      <td>chr17</td>\n",
       "      <td>81174783</td>\n",
       "      <td>81174903</td>\n",
       "      <td>AC139099.1</td>\n",
       "      <td>33812</td>\n",
       "      <td>ENSG00000232938</td>\n",
       "      <td>-0.073778</td>\n",
       "      <td>0.018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638641</th>\n",
       "      <td>chr17</td>\n",
       "      <td>81137640</td>\n",
       "      <td>81137790</td>\n",
       "      <td>1167211</td>\n",
       "      <td>chr17</td>\n",
       "      <td>81174783</td>\n",
       "      <td>81174903</td>\n",
       "      <td>AC139099.1</td>\n",
       "      <td>33812</td>\n",
       "      <td>ENSG00000232938</td>\n",
       "      <td>-0.063751</td>\n",
       "      <td>0.023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638642</th>\n",
       "      <td>chr17</td>\n",
       "      <td>81150865</td>\n",
       "      <td>81151015</td>\n",
       "      <td>1167255</td>\n",
       "      <td>chr17</td>\n",
       "      <td>81174783</td>\n",
       "      <td>81174903</td>\n",
       "      <td>AC139099.1</td>\n",
       "      <td>33812</td>\n",
       "      <td>ENSG00000232938</td>\n",
       "      <td>-0.074219</td>\n",
       "      <td>0.018</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638643</th>\n",
       "      <td>chr17</td>\n",
       "      <td>81151040</td>\n",
       "      <td>81151190</td>\n",
       "      <td>1167256</td>\n",
       "      <td>chr17</td>\n",
       "      <td>81174783</td>\n",
       "      <td>81174903</td>\n",
       "      <td>AC139099.1</td>\n",
       "      <td>33812</td>\n",
       "      <td>ENSG00000232938</td>\n",
       "      <td>-0.089428</td>\n",
       "      <td>0.007</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10832 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       dhs_chr  dhs_start   dhs_end   dhs_id gene_chr  gene_start  gene_end  \\\n",
       "481648   chr17      14460     14610  1051511    chr17       31380     31404   \n",
       "481651   chr17      49180     49330  1051584    chr17       31380     31404   \n",
       "481652   chr17      49800     49950  1051585    chr17       31380     31404   \n",
       "481653   chr17      49980     50130  1051586    chr17       31380     31404   \n",
       "481654   chr17      50160     50310  1051587    chr17       31380     31404   \n",
       "...        ...        ...       ...      ...      ...         ...       ...   \n",
       "424856   chr17   81128545  81128695  1167186    chr17    81037578  81043172   \n",
       "638640   chr17   81137460  81137610  1167210    chr17    81174783  81174903   \n",
       "638641   chr17   81137640  81137790  1167211    chr17    81174783  81174903   \n",
       "638642   chr17   81150865  81151015  1167255    chr17    81174783  81174903   \n",
       "638643   chr17   81151040  81151190  1167256    chr17    81174783  81174903   \n",
       "\n",
       "         gene_name  metaprobeset_id        ensemblID       cor   pval  \n",
       "481648       DOC2B            16908  ENSG00000187939 -0.244659  0.010  \n",
       "481651       DOC2B            16908  ENSG00000187939 -0.242436  0.010  \n",
       "481652       DOC2B            16908  ENSG00000187939 -0.308090  0.003  \n",
       "481653       DOC2B            16908  ENSG00000187939 -0.260791  0.006  \n",
       "481654       DOC2B            16908  ENSG00000187939 -0.287450  0.004  \n",
       "...            ...              ...              ...       ...    ...  \n",
       "424856      METRNL            14362  ENSG00000176845 -0.329763  0.022  \n",
       "638640  AC139099.1            33812  ENSG00000232938 -0.073778  0.018  \n",
       "638641  AC139099.1            33812  ENSG00000232938 -0.063751  0.023  \n",
       "638642  AC139099.1            33812  ENSG00000232938 -0.074219  0.018  \n",
       "638643  AC139099.1            33812  ENSG00000232938 -0.089428  0.007  \n",
       "\n",
       "[10832 rows x 12 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_correlations[sig_correlations[\"dhs_chr\"]==\"chr17\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Grouper for '<class 'pandas.core.frame.DataFrame'>' not 1-dimensional",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\yayag\\Kaggle_SCP\\notebooks\\notebooks_ temp\\ATACseq_preparation.ipynb Cell 21\u001b[0m line \u001b[0;36m1\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X66sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m B_cells_atac\u001b[39m.\u001b[39;49mgroupby(by \u001b[39m=\u001b[39;49m B_cells_atac[[\u001b[39m\"\u001b[39;49m\u001b[39mstart\u001b[39;49m\u001b[39m\"\u001b[39;49m, \u001b[39m\"\u001b[39;49m\u001b[39mend\u001b[39;49m\u001b[39m\"\u001b[39;49m]])\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\frame.py:8872\u001b[0m, in \u001b[0;36mDataFrame.groupby\u001b[1;34m(self, by, axis, level, as_index, sort, group_keys, observed, dropna)\u001b[0m\n\u001b[0;32m   8869\u001b[0m \u001b[39mif\u001b[39;00m level \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m by \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m   8870\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mTypeError\u001b[39;00m(\u001b[39m\"\u001b[39m\u001b[39mYou have to supply one of \u001b[39m\u001b[39m'\u001b[39m\u001b[39mby\u001b[39m\u001b[39m'\u001b[39m\u001b[39m and \u001b[39m\u001b[39m'\u001b[39m\u001b[39mlevel\u001b[39m\u001b[39m'\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m-> 8872\u001b[0m \u001b[39mreturn\u001b[39;00m DataFrameGroupBy(\n\u001b[0;32m   8873\u001b[0m     obj\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m,\n\u001b[0;32m   8874\u001b[0m     keys\u001b[39m=\u001b[39;49mby,\n\u001b[0;32m   8875\u001b[0m     axis\u001b[39m=\u001b[39;49maxis,\n\u001b[0;32m   8876\u001b[0m     level\u001b[39m=\u001b[39;49mlevel,\n\u001b[0;32m   8877\u001b[0m     as_index\u001b[39m=\u001b[39;49mas_index,\n\u001b[0;32m   8878\u001b[0m     sort\u001b[39m=\u001b[39;49msort,\n\u001b[0;32m   8879\u001b[0m     group_keys\u001b[39m=\u001b[39;49mgroup_keys,\n\u001b[0;32m   8880\u001b[0m     observed\u001b[39m=\u001b[39;49mobserved,\n\u001b[0;32m   8881\u001b[0m     dropna\u001b[39m=\u001b[39;49mdropna,\n\u001b[0;32m   8882\u001b[0m )\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\groupby\\groupby.py:1274\u001b[0m, in \u001b[0;36mGroupBy.__init__\u001b[1;34m(self, obj, keys, axis, level, grouper, exclusions, selection, as_index, sort, group_keys, observed, dropna)\u001b[0m\n\u001b[0;32m   1271\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdropna \u001b[39m=\u001b[39m dropna\n\u001b[0;32m   1273\u001b[0m \u001b[39mif\u001b[39;00m grouper \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m-> 1274\u001b[0m     grouper, exclusions, obj \u001b[39m=\u001b[39m get_grouper(\n\u001b[0;32m   1275\u001b[0m         obj,\n\u001b[0;32m   1276\u001b[0m         keys,\n\u001b[0;32m   1277\u001b[0m         axis\u001b[39m=\u001b[39;49maxis,\n\u001b[0;32m   1278\u001b[0m         level\u001b[39m=\u001b[39;49mlevel,\n\u001b[0;32m   1279\u001b[0m         sort\u001b[39m=\u001b[39;49msort,\n\u001b[0;32m   1280\u001b[0m         observed\u001b[39m=\u001b[39;49m\u001b[39mFalse\u001b[39;49;00m \u001b[39mif\u001b[39;49;00m observed \u001b[39mis\u001b[39;49;00m lib\u001b[39m.\u001b[39;49mno_default \u001b[39melse\u001b[39;49;00m observed,\n\u001b[0;32m   1281\u001b[0m         dropna\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdropna,\n\u001b[0;32m   1282\u001b[0m     )\n\u001b[0;32m   1284\u001b[0m \u001b[39mif\u001b[39;00m observed \u001b[39mis\u001b[39;00m lib\u001b[39m.\u001b[39mno_default:\n\u001b[0;32m   1285\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39many\u001b[39m(ping\u001b[39m.\u001b[39m_passed_categorical \u001b[39mfor\u001b[39;00m ping \u001b[39min\u001b[39;00m grouper\u001b[39m.\u001b[39mgroupings):\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\groupby\\grouper.py:1020\u001b[0m, in \u001b[0;36mget_grouper\u001b[1;34m(obj, key, axis, level, sort, observed, validate, dropna)\u001b[0m\n\u001b[0;32m   1015\u001b[0m         in_axis \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m\n\u001b[0;32m   1017\u001b[0m     \u001b[39m# create the Grouping\u001b[39;00m\n\u001b[0;32m   1018\u001b[0m     \u001b[39m# allow us to passing the actual Grouping as the gpr\u001b[39;00m\n\u001b[0;32m   1019\u001b[0m     ping \u001b[39m=\u001b[39m (\n\u001b[1;32m-> 1020\u001b[0m         Grouping(\n\u001b[0;32m   1021\u001b[0m             group_axis,\n\u001b[0;32m   1022\u001b[0m             gpr,\n\u001b[0;32m   1023\u001b[0m             obj\u001b[39m=\u001b[39;49mobj,\n\u001b[0;32m   1024\u001b[0m             level\u001b[39m=\u001b[39;49mlevel,\n\u001b[0;32m   1025\u001b[0m             sort\u001b[39m=\u001b[39;49msort,\n\u001b[0;32m   1026\u001b[0m             observed\u001b[39m=\u001b[39;49mobserved,\n\u001b[0;32m   1027\u001b[0m             in_axis\u001b[39m=\u001b[39;49min_axis,\n\u001b[0;32m   1028\u001b[0m             dropna\u001b[39m=\u001b[39;49mdropna,\n\u001b[0;32m   1029\u001b[0m         )\n\u001b[0;32m   1030\u001b[0m         \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(gpr, Grouping)\n\u001b[0;32m   1031\u001b[0m         \u001b[39melse\u001b[39;00m gpr\n\u001b[0;32m   1032\u001b[0m     )\n\u001b[0;32m   1034\u001b[0m     groupings\u001b[39m.\u001b[39mappend(ping)\n\u001b[0;32m   1036\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(groupings) \u001b[39m==\u001b[39m \u001b[39m0\u001b[39m \u001b[39mand\u001b[39;00m \u001b[39mlen\u001b[39m(obj):\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\groupby\\grouper.py:601\u001b[0m, in \u001b[0;36mGrouping.__init__\u001b[1;34m(self, index, grouper, obj, level, sort, observed, in_axis, dropna, uniques)\u001b[0m\n\u001b[0;32m    599\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mgetattr\u001b[39m(grouping_vector, \u001b[39m\"\u001b[39m\u001b[39mndim\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m1\u001b[39m) \u001b[39m!=\u001b[39m \u001b[39m1\u001b[39m:\n\u001b[0;32m    600\u001b[0m     t \u001b[39m=\u001b[39m \u001b[39mstr\u001b[39m(\u001b[39mtype\u001b[39m(grouping_vector))\n\u001b[1;32m--> 601\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mGrouper for \u001b[39m\u001b[39m'\u001b[39m\u001b[39m{\u001b[39;00mt\u001b[39m}\u001b[39;00m\u001b[39m'\u001b[39m\u001b[39m not 1-dimensional\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    603\u001b[0m grouping_vector \u001b[39m=\u001b[39m index\u001b[39m.\u001b[39mmap(grouping_vector)\n\u001b[0;32m    605\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m (\n\u001b[0;32m    606\u001b[0m     \u001b[39mhasattr\u001b[39m(grouping_vector, \u001b[39m\"\u001b[39m\u001b[39m__len__\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[0;32m    607\u001b[0m     \u001b[39mand\u001b[39;00m \u001b[39mlen\u001b[39m(grouping_vector) \u001b[39m==\u001b[39m \u001b[39mlen\u001b[39m(index)\n\u001b[0;32m    608\u001b[0m ):\n",
      "\u001b[1;31mValueError\u001b[0m: Grouper for '<class 'pandas.core.frame.DataFrame'>' not 1-dimensional"
     ]
    }
   ],
   "source": [
    "def create_gene_atac4(atac, cor):\n",
    "    gene_atac = pd.DataFrame()\n",
    "    all_cor = cor.copy()\n",
    "    all_atac = atac.copy()\n",
    "    \n",
    "    for cn, chromosome in enumerate(all_cor[\"dhs_chr\"].unique()):\n",
    "        atac = all_atac[all_atac[\"chromosome\"]==chromosome]\n",
    "        cor = all_cor[all_cor[\"dhs_chr\"]==chromosome]\n",
    "        print(f\"chromosome no: {cn} | name: {chromosome}\")\n",
    "        print(f\"atac len for chromosome: {len(atac)}\")\n",
    "        print(f\"cor len for chromosome: {len(cor)}\")\n",
    "\n",
    "        p_atac, p_cor = 0, 0\n",
    "\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "B_cells_sig_start = np.array(sig_correlations[\"gene_start\"]).reshape(1, -1)\n",
    "B_cells_sig_end = np.array(sig_correlations[\"gene_end\"]).reshape(1, -1)\n",
    "B_cells_atac_start = np.array(B_cells_atac[\"start\"]).reshape(-1, 1)\n",
    "B_cells_atac_end = np.array(B_cells_atac[\"end\"]).reshape(-1, 1)\n",
    "B_cells_atac_Ncount = np.array(B_cells_atac[\"normalized_count\"]).reshape(-1, 1)\n",
    "sig_genes = sig_correlations[\"ensemblID\"].to_list()\n",
    "# B_cells_broadcast = np.broadcast(sig_array, B_cells_atac_array[:len(sig_array)])\n",
    "# out = np.empty(B_cells_broadcast.shape)\n",
    "#out.flat = [min(int(sig_end), int(atac_end))-max(int(sig_start), int(atac_start)+1) for ((sig_start, sig_end), (atac_start, atac_end)) in B_cells_broadcast]\n",
    "#out"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compute_sig_access(sig_start, sig_end, atac_starts, atac_ends, atac_Ncounts):\n",
    "    overlap = (np.minimum(sig_end, atac_ends) - np.maximum(sig_start, atac_starts))/(sig_end-sig_start)\n",
    "    overlap[overlap<0]=0\n",
    "    s = overlap.sum()\n",
    "    if s > 0:\n",
    "        accessScore = (overlap * atac_Ncounts).sum() / overlap.sum()\n",
    "    else:\n",
    "        accessScore = -1\n",
    "         \n",
    "    return accessScore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import temp\n",
    "from temp import test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/177 [00:00<?, ?it/s]"
     ]
    }
   ],
   "source": [
    "from multiprocessing import Pool\n",
    "from temp import test\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "\n",
    "    def get_gene_accessibility(sig_starts, sig_ends, atac_starts, atac_ends, atac_Ncounts, sig_genes): \n",
    "        gene_accessibility = pd.DataFrame(columns=[\"ensembleID\", \"accessibility\"])\n",
    "        \n",
    "        batch_size = 1000\n",
    "        batch_num = (sig_starts.shape[1] // batch_size) + 1\n",
    "        for i in tqdm(range(batch_num)):\n",
    "            if i < batch_num - 1:\n",
    "                sig_start = sig_starts[0, i*batch_size:(i+1)*batch_size]\n",
    "                sig_end = sig_ends[0, i*batch_size:(i+1)*batch_size]\n",
    "            else:\n",
    "                sig_start = sig_starts[0, i*batch_size:]\n",
    "                sig_end = sig_ends[0, i*batch_size:]\n",
    "            \n",
    "            c = test(atac_starts, atac_ends, atac_Ncounts)\n",
    "            p = Pool(16)\n",
    "            access = p.starmap(c.compute_sig_access, zip(list(sig_start), list(sig_end)))\n",
    "            gene = sig_genes[i]\n",
    "            gene_accessibility = pd.concat([gene_accessibility, pd.DataFrame({\"gene\":gene, \"accessibility\":access})])\n",
    "        return gene_accessibility\n",
    "    \n",
    "    B_cells_gene_accessibility = get_gene_accessibility(B_cells_sig_start, B_cells_sig_end, B_cells_atac_start, B_cells_atac_end, B_cells_atac_Ncount, sig_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from multiprocessing import Pool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "from temp import f\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    p = Pool(3)\n",
    "    params = [(1, 2), (2, 3), (3, 4)]\n",
    "    results = p.starmap(f, params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 6, 12]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "results[0].get()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "d.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/8818 [00:00<?, ?it/s]C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\2134768079.py:4: RuntimeWarning: invalid value encountered in divide\n",
      "  accessScore = (overlap * atac_Ncounts).sum(0) / overlap.sum(0)\n",
      "C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\470227413.py:15: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  gene_accessibility = pd.concat([gene_accessibility, pd.DataFrame({\"gene\":gene, \"accessibility\":access})])\n",
      "  0%|          | 1/8818 [00:49<121:36:53, 49.66s/it]C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\2134768079.py:4: RuntimeWarning: invalid value encountered in divide\n",
      "  accessScore = (overlap * atac_Ncounts).sum(0) / overlap.sum(0)\n",
      "  0%|          | 2/8818 [01:08<77:05:10, 31.48s/it] C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\2134768079.py:4: RuntimeWarning: invalid value encountered in divide\n",
      "  accessScore = (overlap * atac_Ncounts).sum(0) / overlap.sum(0)\n",
      "  0%|          | 3/8818 [01:26<62:11:11, 25.40s/it]C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\2134768079.py:4: RuntimeWarning: invalid value encountered in divide\n",
      "  accessScore = (overlap * atac_Ncounts).sum(0) / overlap.sum(0)\n",
      "  0%|          | 10/8818 [03:28<45:18:34, 18.52s/it]C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\2134768079.py:4: RuntimeWarning: invalid value encountered in divide\n",
      "  accessScore = (overlap * atac_Ncounts).sum(0) / overlap.sum(0)\n",
      "  0%|          | 11/8818 [03:48<46:23:28, 18.96s/it]C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\2134768079.py:4: RuntimeWarning: invalid value encountered in divide\n",
      "  accessScore = (overlap * atac_Ncounts).sum(0) / overlap.sum(0)\n",
      "  0%|          | 12/8818 [04:08<47:17:02, 19.33s/it]C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\2134768079.py:4: RuntimeWarning: invalid value encountered in divide\n",
      "  accessScore = (overlap * atac_Ncounts).sum(0) / overlap.sum(0)\n",
      "  0%|          | 13/8818 [04:27<46:56:42, 19.19s/it]C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_19228\\2134768079.py:4: RuntimeWarning: invalid value encountered in divide\n",
      "  accessScore = (overlap * atac_Ncounts).sum(0) / overlap.sum(0)\n",
      "  0%|          | 14/8818 [05:01<52:35:49, 21.51s/it]\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\yayag\\Kaggle_SCP\\notebooks\\notebooks_ temp\\ATACseq_preparation.ipynb Cell 24\u001b[0m line \u001b[0;36m1\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m B_cells_gene_accessibility \u001b[39m=\u001b[39m get_gene_accessibility(B_cells_sig_start, B_cells_sig_end, B_cells_atac_start, B_cells_atac_end, B_cells_atac_Ncount, sig_genes)\n",
      "\u001b[1;32mc:\\Users\\yayag\\Kaggle_SCP\\notebooks\\notebooks_ temp\\ATACseq_preparation.ipynb Cell 24\u001b[0m line \u001b[0;36m1\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=10'>11</a>\u001b[0m     sig_start \u001b[39m=\u001b[39m sig_starts[\u001b[39m0\u001b[39m, i\u001b[39m*\u001b[39mbatch_size:]\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=11'>12</a>\u001b[0m     sig_end \u001b[39m=\u001b[39m sig_ends[\u001b[39m0\u001b[39m, i\u001b[39m*\u001b[39mbatch_size:]\n\u001b[1;32m---> <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=12'>13</a>\u001b[0m access \u001b[39m=\u001b[39m compute_sig_access(sig_start, sig_end, atac_starts, atac_ends, atac_Ncounts)\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=13'>14</a>\u001b[0m gene \u001b[39m=\u001b[39m sig_genes[i]\n\u001b[0;32m     <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=14'>15</a>\u001b[0m gene_accessibility \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39mconcat([gene_accessibility, pd\u001b[39m.\u001b[39mDataFrame({\u001b[39m\"\u001b[39m\u001b[39mgene\u001b[39m\u001b[39m\"\u001b[39m:gene, \u001b[39m\"\u001b[39m\u001b[39maccessibility\u001b[39m\u001b[39m\"\u001b[39m:access})])\n",
      "\u001b[1;32mc:\\Users\\yayag\\Kaggle_SCP\\notebooks\\notebooks_ temp\\ATACseq_preparation.ipynb Cell 24\u001b[0m line \u001b[0;36m4\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m overlap \u001b[39m=\u001b[39m (np\u001b[39m.\u001b[39mminimum(sig_end, atac_ends) \u001b[39m-\u001b[39m np\u001b[39m.\u001b[39mmaximum(sig_start, atac_starts))\u001b[39m/\u001b[39m(sig_end\u001b[39m-\u001b[39msig_start)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m overlap[overlap\u001b[39m<\u001b[39m\u001b[39m0\u001b[39m]\u001b[39m=\u001b[39m\u001b[39m0\u001b[39m\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m accessScore \u001b[39m=\u001b[39m (overlap \u001b[39m*\u001b[39;49m atac_Ncounts)\u001b[39m.\u001b[39;49msum(\u001b[39m0\u001b[39;49m) \u001b[39m/\u001b[39m overlap\u001b[39m.\u001b[39msum(\u001b[39m0\u001b[39m)\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X56sZmlsZQ%3D%3D?line=5'>6</a>\u001b[0m \u001b[39mreturn\u001b[39;00m accessScore\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\numpy\\core\\_methods.py:47\u001b[0m, in \u001b[0;36m_sum\u001b[1;34m(a, axis, dtype, out, keepdims, initial, where)\u001b[0m\n\u001b[0;32m     43\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_amin\u001b[39m(a, axis\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, out\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, keepdims\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[0;32m     44\u001b[0m           initial\u001b[39m=\u001b[39m_NoValue, where\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m):\n\u001b[0;32m     45\u001b[0m     \u001b[39mreturn\u001b[39;00m umr_minimum(a, axis, \u001b[39mNone\u001b[39;00m, out, keepdims, initial, where)\n\u001b[1;32m---> 47\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_sum\u001b[39m(a, axis\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, dtype\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, out\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, keepdims\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[0;32m     48\u001b[0m          initial\u001b[39m=\u001b[39m_NoValue, where\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m):\n\u001b[0;32m     49\u001b[0m     \u001b[39mreturn\u001b[39;00m umr_sum(a, axis, dtype, out, keepdims, initial, where)\n\u001b[0;32m     51\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_prod\u001b[39m(a, axis\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, dtype\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, out\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, keepdims\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m,\n\u001b[0;32m     52\u001b[0m           initial\u001b[39m=\u001b[39m_NoValue, where\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m):\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "B_cells_gene_accessibility = get_gene_accessibility(B_cells_sig_start, B_cells_sig_end, B_cells_atac_start, B_cells_atac_end, B_cells_atac_Ncount, sig_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'ENSG00000232325'"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_genes[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.4062807866574807"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sth[sth<0]=0\n",
    "(sth * B_cells_atac_Ncount).sum() / sth.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "x, y = sig_array[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71699"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "x"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>location</th>\n",
       "      <th>count</th>\n",
       "      <th>normalized_count</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>162859353</th>\n",
       "      <td>f03f7ded47ff1b35</td>\n",
       "      <td>chr2:9976-10826</td>\n",
       "      <td>2</td>\n",
       "      <td>3.021235</td>\n",
       "      <td>chr2</td>\n",
       "      <td>9976</td>\n",
       "      <td>10826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>126050035</th>\n",
       "      <td>b838a74cdfdb1634</td>\n",
       "      <td>chr2:9976-10826</td>\n",
       "      <td>2</td>\n",
       "      <td>2.050489</td>\n",
       "      <td>chr2</td>\n",
       "      <td>9976</td>\n",
       "      <td>10826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>116961257</th>\n",
       "      <td>aa74fcd6c46bd74e</td>\n",
       "      <td>chr2:9976-10826</td>\n",
       "      <td>2</td>\n",
       "      <td>1.265548</td>\n",
       "      <td>chr2</td>\n",
       "      <td>9976</td>\n",
       "      <td>10826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>163146930</th>\n",
       "      <td>f0baf396e3ab61a5</td>\n",
       "      <td>chr2:9976-10826</td>\n",
       "      <td>2</td>\n",
       "      <td>5.448502</td>\n",
       "      <td>chr2</td>\n",
       "      <td>9976</td>\n",
       "      <td>10826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146133662</th>\n",
       "      <td>d7352ad054c67b77</td>\n",
       "      <td>chr2:9976-10826</td>\n",
       "      <td>2</td>\n",
       "      <td>6.470487</td>\n",
       "      <td>chr2</td>\n",
       "      <td>9976</td>\n",
       "      <td>10826</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49652552</th>\n",
       "      <td>47eb9c3e01c6a601</td>\n",
       "      <td>chr2:242182974-242183864</td>\n",
       "      <td>2</td>\n",
       "      <td>4.400481</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242182974</td>\n",
       "      <td>242183864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13266114</th>\n",
       "      <td>12c2025cf0abd959</td>\n",
       "      <td>chr2:242182974-242183864</td>\n",
       "      <td>2</td>\n",
       "      <td>2.790936</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242182974</td>\n",
       "      <td>242183864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23090053</th>\n",
       "      <td>2181f8df45af4076</td>\n",
       "      <td>chr2:242182974-242183864</td>\n",
       "      <td>2</td>\n",
       "      <td>1.426278</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242182974</td>\n",
       "      <td>242183864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105988782</th>\n",
       "      <td>9af75c60c38378f5</td>\n",
       "      <td>chr2:242182974-242183864</td>\n",
       "      <td>2</td>\n",
       "      <td>1.777322</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242182974</td>\n",
       "      <td>242183864</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13350981</th>\n",
       "      <td>12e6deb8dec88770</td>\n",
       "      <td>chr2:242182974-242183864</td>\n",
       "      <td>2</td>\n",
       "      <td>2.799781</td>\n",
       "      <td>chr2</td>\n",
       "      <td>242182974</td>\n",
       "      <td>242183864</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1997658 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     obs_id                  location  count  \\\n",
       "162859353  f03f7ded47ff1b35           chr2:9976-10826      2   \n",
       "126050035  b838a74cdfdb1634           chr2:9976-10826      2   \n",
       "116961257  aa74fcd6c46bd74e           chr2:9976-10826      2   \n",
       "163146930  f0baf396e3ab61a5           chr2:9976-10826      2   \n",
       "146133662  d7352ad054c67b77           chr2:9976-10826      2   \n",
       "...                     ...                       ...    ...   \n",
       "49652552   47eb9c3e01c6a601  chr2:242182974-242183864      2   \n",
       "13266114   12c2025cf0abd959  chr2:242182974-242183864      2   \n",
       "23090053   2181f8df45af4076  chr2:242182974-242183864      2   \n",
       "105988782  9af75c60c38378f5  chr2:242182974-242183864      2   \n",
       "13350981   12e6deb8dec88770  chr2:242182974-242183864      2   \n",
       "\n",
       "           normalized_count chromosome      start        end  \n",
       "162859353          3.021235       chr2       9976      10826  \n",
       "126050035          2.050489       chr2       9976      10826  \n",
       "116961257          1.265548       chr2       9976      10826  \n",
       "163146930          5.448502       chr2       9976      10826  \n",
       "146133662          6.470487       chr2       9976      10826  \n",
       "...                     ...        ...        ...        ...  \n",
       "49652552           4.400481       chr2  242182974  242183864  \n",
       "13266114           2.790936       chr2  242182974  242183864  \n",
       "23090053           1.426278       chr2  242182974  242183864  \n",
       "105988782          1.777322       chr2  242182974  242183864  \n",
       "13350981           2.799781       chr2  242182974  242183864  \n",
       "\n",
       "[1997658 rows x 7 columns]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "B_cells_atac[B_cells_atac[\"chromosome\"]==\"chr2\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>ensembleID</th>\n",
       "      <th>intersection</th>\n",
       "      <th>cor_start</th>\n",
       "      <th>cor_end</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>841a3e50861e4ab1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>3.898647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4240bac01e20b43e</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>1.994656</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>444903c060e9f7af</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>4.962905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>fec6b653a30382fc</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>4.186073</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>7e37aae9075ffeda</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>2.736343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>b88c4f893601c319</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>1.371382</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ea503ec86c1aa28a</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>2.176123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>e35b3f5df751962b</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>2.187354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>98c7a5622654d820</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>3.953332</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              obs_id chromosome       ensembleID  intersection  cor_start  \\\n",
       "0   841a3e50861e4ab1       chr7  ENSG00000232325           1.0    10080.0   \n",
       "2   4240bac01e20b43e       chr7  ENSG00000232325           1.0    10080.0   \n",
       "4   444903c060e9f7af       chr7  ENSG00000232325           1.0    10080.0   \n",
       "6   fec6b653a30382fc       chr7  ENSG00000232325           1.0    10080.0   \n",
       "8   7e37aae9075ffeda       chr7  ENSG00000232325           1.0    10080.0   \n",
       "10  b88c4f893601c319       chr7  ENSG00000232325           1.0    10080.0   \n",
       "12  ea503ec86c1aa28a       chr7  ENSG00000232325           1.0    10080.0   \n",
       "14  e35b3f5df751962b       chr7  ENSG00000232325           1.0    10080.0   \n",
       "16  98c7a5622654d820       chr7  ENSG00000232325           1.0    10080.0   \n",
       "\n",
       "    cor_end     score  \n",
       "0   10230.0  3.898647  \n",
       "2   10230.0  1.994656  \n",
       "4   10230.0  4.962905  \n",
       "6   10230.0  4.186073  \n",
       "8   10230.0  2.736343  \n",
       "10  10230.0  1.371382  \n",
       "12  10230.0  2.176123  \n",
       "14  10230.0  2.187354  \n",
       "16  10230.0  3.953332  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "B_cells_gene_atac[B_cells_gene_atac[\"chromosome\"]!=0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10230"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sig_correlations[\"dhs_end\"].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>ensembleID</th>\n",
       "      <th>intersection</th>\n",
       "      <th>cor_start</th>\n",
       "      <th>cor_end</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>841a3e50861e4ab1</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>3.898647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3dc613637024e42b</td>\n",
       "      <td>chr7</td>\n",
       "      <td>ENSG00000232325</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10080.0</td>\n",
       "      <td>10230.0</td>\n",
       "      <td>2.142502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176347</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176348</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176349</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176350</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>176351</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>176352 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  obs_id chromosome       ensembleID  intersection  cor_start  \\\n",
       "0       841a3e50861e4ab1       chr7  ENSG00000232325           1.0    10080.0   \n",
       "1                    0.0        0.0              0.0           0.0        0.0   \n",
       "2       3dc613637024e42b       chr7  ENSG00000232325           1.0    10080.0   \n",
       "3                    0.0        0.0              0.0           0.0        0.0   \n",
       "4                    0.0        0.0              0.0           0.0        0.0   \n",
       "...                  ...        ...              ...           ...        ...   \n",
       "176347               0.0        0.0              0.0           0.0        0.0   \n",
       "176348               0.0        0.0              0.0           0.0        0.0   \n",
       "176349               0.0        0.0              0.0           0.0        0.0   \n",
       "176350               0.0        0.0              0.0           0.0        0.0   \n",
       "176351               0.0        0.0              0.0           0.0        0.0   \n",
       "\n",
       "        cor_end     score  \n",
       "0       10230.0  3.898647  \n",
       "1           0.0  0.000000  \n",
       "2       10230.0  2.142502  \n",
       "3           0.0  0.000000  \n",
       "4           0.0  0.000000  \n",
       "...         ...       ...  \n",
       "176347      0.0  0.000000  \n",
       "176348      0.0  0.000000  \n",
       "176349      0.0  0.000000  \n",
       "176350      0.0  0.000000  \n",
       "176351      0.0  0.000000  \n",
       "\n",
       "[176352 rows x 7 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "B_cells_gene_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(set(range(1,11)).intersection(set(range(10,15))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "'RP11' is not in list",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\yayag\\Kaggle_SCP\\notebooks\\notebooks_ temp\\ATACseq_preparation.ipynb Cell 10\u001b[0m line \u001b[0;36m1\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#X22sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m gene_names\u001b[39m.\u001b[39;49mindex(\u001b[39m\"\u001b[39;49m\u001b[39mRP11\u001b[39;49m\u001b[39m\"\u001b[39;49m)\n",
      "\u001b[1;31mValueError\u001b[0m: 'RP11' is not in list"
     ]
    }
   ],
   "source": [
    "gene_names.index(\"RP11\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>location</th>\n",
       "      <th>count</th>\n",
       "      <th>normalized_count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>258406</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:100006075-100006963</td>\n",
       "      <td>2</td>\n",
       "      <td>2.479890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258407</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:100229082-100229931</td>\n",
       "      <td>1</td>\n",
       "      <td>0.824665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258408</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:100286187-100287039</td>\n",
       "      <td>2</td>\n",
       "      <td>0.984346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258409</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:100339340-100339978</td>\n",
       "      <td>2</td>\n",
       "      <td>3.327292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258410</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:101066169-101066970</td>\n",
       "      <td>4</td>\n",
       "      <td>3.590057</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251363</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:19499847-19500746</td>\n",
       "      <td>1</td>\n",
       "      <td>3.342186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251364</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:19566990-19567754</td>\n",
       "      <td>1</td>\n",
       "      <td>1.154666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251365</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:20735890-20736626</td>\n",
       "      <td>3</td>\n",
       "      <td>7.049924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251366</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:21260054-21260842</td>\n",
       "      <td>2</td>\n",
       "      <td>3.967924</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251367</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:7273431-7274301</td>\n",
       "      <td>2</td>\n",
       "      <td>1.250685</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215992962 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     obs_id                   location  count  \\\n",
       "258406     004dd1a01106608a  chr10:100006075-100006963      2   \n",
       "258407     004dd1a01106608a  chr10:100229082-100229931      1   \n",
       "258408     004dd1a01106608a  chr10:100286187-100287039      2   \n",
       "258409     004dd1a01106608a  chr10:100339340-100339978      2   \n",
       "258410     004dd1a01106608a  chr10:101066169-101066970      4   \n",
       "...                     ...                        ...    ...   \n",
       "216251363  ffffeba9dbf2bab1     chrY:19499847-19500746      1   \n",
       "216251364  ffffeba9dbf2bab1     chrY:19566990-19567754      1   \n",
       "216251365  ffffeba9dbf2bab1     chrY:20735890-20736626      3   \n",
       "216251366  ffffeba9dbf2bab1     chrY:21260054-21260842      2   \n",
       "216251367  ffffeba9dbf2bab1       chrY:7273431-7274301      2   \n",
       "\n",
       "           normalized_count  \n",
       "258406             2.479890  \n",
       "258407             0.824665  \n",
       "258408             0.984346  \n",
       "258409             3.327292  \n",
       "258410             3.590057  \n",
       "...                     ...  \n",
       "216251363          3.342186  \n",
       "216251364          1.154666  \n",
       "216251365          7.049924  \n",
       "216251366          3.967924  \n",
       "216251367          1.250685  \n",
       "\n",
       "[215992962 rows x 4 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac = multiome_df.loc[258406:]\n",
    "atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_9316\\820159844.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  atac[\"chromosome\"] = atac[\"location\"].apply(get_chr)\n"
     ]
    }
   ],
   "source": [
    "def get_chr(all_str):\n",
    "    index = all_str.find(\":\") \n",
    "    return all_str[:index] \n",
    "\n",
    "\n",
    "atac[\"chromosome\"] = atac[\"location\"].apply(get_chr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_9316\\2006402745.py:5: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  atac[\"start\"] = atac[\"location\"].apply(get_start)\n"
     ]
    }
   ],
   "source": [
    "def get_start(all_str):\n",
    "    return all_str[all_str.find(\":\")+1:all_str.find(\"-\")] \n",
    "\n",
    "\n",
    "atac[\"start\"] = atac[\"location\"].apply(get_start)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32mc:\\Users\\yayag\\Kaggle_SCP\\notebooks\\notebooks_ temp\\ATACseq_preparation.ipynb Cell 6\u001b[0m line \u001b[0;36m5\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#W5sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_end\u001b[39m(all_str):\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#W5sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m all_str[all_str\u001b[39m.\u001b[39mfind(\u001b[39m\"\u001b[39m\u001b[39m-\u001b[39m\u001b[39m\"\u001b[39m)\u001b[39m+\u001b[39m\u001b[39m1\u001b[39m:] \n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#W5sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m atac[\u001b[39m\"\u001b[39m\u001b[39mend\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m atac[\u001b[39m\"\u001b[39;49m\u001b[39mlocation\u001b[39;49m\u001b[39m\"\u001b[39;49m]\u001b[39m.\u001b[39;49mapply(get_end)\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\series.py:4760\u001b[0m, in \u001b[0;36mSeries.apply\u001b[1;34m(self, func, convert_dtype, args, by_row, **kwargs)\u001b[0m\n\u001b[0;32m   4625\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mapply\u001b[39m(\n\u001b[0;32m   4626\u001b[0m     \u001b[39mself\u001b[39m,\n\u001b[0;32m   4627\u001b[0m     func: AggFuncType,\n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   4632\u001b[0m     \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs,\n\u001b[0;32m   4633\u001b[0m ) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m DataFrame \u001b[39m|\u001b[39m Series:\n\u001b[0;32m   4634\u001b[0m \u001b[39m    \u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[0;32m   4635\u001b[0m \u001b[39m    Invoke function on values of Series.\u001b[39;00m\n\u001b[0;32m   4636\u001b[0m \n\u001b[1;32m   (...)\u001b[0m\n\u001b[0;32m   4751\u001b[0m \u001b[39m    dtype: float64\u001b[39;00m\n\u001b[0;32m   4752\u001b[0m \u001b[39m    \"\"\"\u001b[39;00m\n\u001b[0;32m   4753\u001b[0m     \u001b[39mreturn\u001b[39;00m SeriesApply(\n\u001b[0;32m   4754\u001b[0m         \u001b[39mself\u001b[39;49m,\n\u001b[0;32m   4755\u001b[0m         func,\n\u001b[0;32m   4756\u001b[0m         convert_dtype\u001b[39m=\u001b[39;49mconvert_dtype,\n\u001b[0;32m   4757\u001b[0m         by_row\u001b[39m=\u001b[39;49mby_row,\n\u001b[0;32m   4758\u001b[0m         args\u001b[39m=\u001b[39;49margs,\n\u001b[0;32m   4759\u001b[0m         kwargs\u001b[39m=\u001b[39;49mkwargs,\n\u001b[1;32m-> 4760\u001b[0m     )\u001b[39m.\u001b[39;49mapply()\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\apply.py:1207\u001b[0m, in \u001b[0;36mSeriesApply.apply\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1204\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mapply_compat()\n\u001b[0;32m   1206\u001b[0m \u001b[39m# self.func is Callable\u001b[39;00m\n\u001b[1;32m-> 1207\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mapply_standard()\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\apply.py:1287\u001b[0m, in \u001b[0;36mSeriesApply.apply_standard\u001b[1;34m(self)\u001b[0m\n\u001b[0;32m   1281\u001b[0m \u001b[39m# row-wise access\u001b[39;00m\n\u001b[0;32m   1282\u001b[0m \u001b[39m# apply doesn't have a `na_action` keyword and for backward compat reasons\u001b[39;00m\n\u001b[0;32m   1283\u001b[0m \u001b[39m# we need to give `na_action=\"ignore\"` for categorical data.\u001b[39;00m\n\u001b[0;32m   1284\u001b[0m \u001b[39m# TODO: remove the `na_action=\"ignore\"` when that default has been changed in\u001b[39;00m\n\u001b[0;32m   1285\u001b[0m \u001b[39m#  Categorical (GH51645).\u001b[39;00m\n\u001b[0;32m   1286\u001b[0m action \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mignore\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(obj\u001b[39m.\u001b[39mdtype, CategoricalDtype) \u001b[39melse\u001b[39;00m \u001b[39mNone\u001b[39;00m\n\u001b[1;32m-> 1287\u001b[0m mapped \u001b[39m=\u001b[39m obj\u001b[39m.\u001b[39;49m_map_values(\n\u001b[0;32m   1288\u001b[0m     mapper\u001b[39m=\u001b[39;49mcurried, na_action\u001b[39m=\u001b[39;49maction, convert\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mconvert_dtype\n\u001b[0;32m   1289\u001b[0m )\n\u001b[0;32m   1291\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mlen\u001b[39m(mapped) \u001b[39mand\u001b[39;00m \u001b[39misinstance\u001b[39m(mapped[\u001b[39m0\u001b[39m], ABCSeries):\n\u001b[0;32m   1292\u001b[0m     \u001b[39m# GH#43986 Need to do list(mapped) in order to get treated as nested\u001b[39;00m\n\u001b[0;32m   1293\u001b[0m     \u001b[39m#  See also GH#25959 regarding EA support\u001b[39;00m\n\u001b[0;32m   1294\u001b[0m     \u001b[39mreturn\u001b[39;00m obj\u001b[39m.\u001b[39m_constructor_expanddim(\u001b[39mlist\u001b[39m(mapped), index\u001b[39m=\u001b[39mobj\u001b[39m.\u001b[39mindex)\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\base.py:921\u001b[0m, in \u001b[0;36mIndexOpsMixin._map_values\u001b[1;34m(self, mapper, na_action, convert)\u001b[0m\n\u001b[0;32m    918\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(arr, ExtensionArray):\n\u001b[0;32m    919\u001b[0m     \u001b[39mreturn\u001b[39;00m arr\u001b[39m.\u001b[39mmap(mapper, na_action\u001b[39m=\u001b[39mna_action)\n\u001b[1;32m--> 921\u001b[0m \u001b[39mreturn\u001b[39;00m algorithms\u001b[39m.\u001b[39;49mmap_array(arr, mapper, na_action\u001b[39m=\u001b[39;49mna_action, convert\u001b[39m=\u001b[39;49mconvert)\n",
      "File \u001b[1;32mc:\\Users\\yayag\\anaconda3\\envs\\kaggle_scp\\lib\\site-packages\\pandas\\core\\algorithms.py:1814\u001b[0m, in \u001b[0;36mmap_array\u001b[1;34m(arr, mapper, na_action, convert)\u001b[0m\n\u001b[0;32m   1812\u001b[0m values \u001b[39m=\u001b[39m arr\u001b[39m.\u001b[39mastype(\u001b[39mobject\u001b[39m, copy\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m)\n\u001b[0;32m   1813\u001b[0m \u001b[39mif\u001b[39;00m na_action \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m-> 1814\u001b[0m     \u001b[39mreturn\u001b[39;00m lib\u001b[39m.\u001b[39;49mmap_infer(values, mapper, convert\u001b[39m=\u001b[39;49mconvert)\n\u001b[0;32m   1815\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m   1816\u001b[0m     \u001b[39mreturn\u001b[39;00m lib\u001b[39m.\u001b[39mmap_infer_mask(\n\u001b[0;32m   1817\u001b[0m         values, mapper, mask\u001b[39m=\u001b[39misna(values)\u001b[39m.\u001b[39mview(np\u001b[39m.\u001b[39muint8), convert\u001b[39m=\u001b[39mconvert\n\u001b[0;32m   1818\u001b[0m     )\n",
      "File \u001b[1;32mlib.pyx:2917\u001b[0m, in \u001b[0;36mpandas._libs.lib.map_infer\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;32mc:\\Users\\yayag\\Kaggle_SCP\\notebooks\\notebooks_ temp\\ATACseq_preparation.ipynb Cell 6\u001b[0m line \u001b[0;36m1\n\u001b[1;32m----> <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#W5sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mget_end\u001b[39m(all_str):\n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#W5sZmlsZQ%3D%3D?line=1'>2</a>\u001b[0m     \u001b[39mreturn\u001b[39;00m all_str[all_str\u001b[39m.\u001b[39mfind(\u001b[39m\"\u001b[39m\u001b[39m-\u001b[39m\u001b[39m\"\u001b[39m)\u001b[39m+\u001b[39m\u001b[39m1\u001b[39m:] \n\u001b[0;32m      <a href='vscode-notebook-cell:/c%3A/Users/yayag/Kaggle_SCP/notebooks/notebooks_%20temp/ATACseq_preparation.ipynb#W5sZmlsZQ%3D%3D?line=4'>5</a>\u001b[0m atac[\u001b[39m\"\u001b[39m\u001b[39mend\u001b[39m\u001b[39m\"\u001b[39m] \u001b[39m=\u001b[39m atac[\u001b[39m\"\u001b[39m\u001b[39mlocation\u001b[39m\u001b[39m\"\u001b[39m]\u001b[39m.\u001b[39mapply(get_end)\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\yayag\\AppData\\Local\\Temp\\ipykernel_9316\\905927124.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  atac.drop(\"count\", axis=1, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "atac.drop(\"count\", axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>obs_id</th>\n",
       "      <th>location</th>\n",
       "      <th>normalized_count</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>start</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>258406</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:100006075-100006963</td>\n",
       "      <td>2.479890</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100006075</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258407</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:100229082-100229931</td>\n",
       "      <td>0.824665</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100229082</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258408</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:100286187-100287039</td>\n",
       "      <td>0.984346</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100286187</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258409</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:100339340-100339978</td>\n",
       "      <td>3.327292</td>\n",
       "      <td>chr10</td>\n",
       "      <td>100339340</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>258410</th>\n",
       "      <td>004dd1a01106608a</td>\n",
       "      <td>chr10:101066169-101066970</td>\n",
       "      <td>3.590057</td>\n",
       "      <td>chr10</td>\n",
       "      <td>101066169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251363</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:19499847-19500746</td>\n",
       "      <td>3.342186</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19499847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251364</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:19566990-19567754</td>\n",
       "      <td>1.154666</td>\n",
       "      <td>chrY</td>\n",
       "      <td>19566990</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251365</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:20735890-20736626</td>\n",
       "      <td>7.049924</td>\n",
       "      <td>chrY</td>\n",
       "      <td>20735890</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251366</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:21260054-21260842</td>\n",
       "      <td>3.967924</td>\n",
       "      <td>chrY</td>\n",
       "      <td>21260054</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>216251367</th>\n",
       "      <td>ffffeba9dbf2bab1</td>\n",
       "      <td>chrY:7273431-7274301</td>\n",
       "      <td>1.250685</td>\n",
       "      <td>chrY</td>\n",
       "      <td>7273431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>215992962 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     obs_id                   location  normalized_count  \\\n",
       "258406     004dd1a01106608a  chr10:100006075-100006963          2.479890   \n",
       "258407     004dd1a01106608a  chr10:100229082-100229931          0.824665   \n",
       "258408     004dd1a01106608a  chr10:100286187-100287039          0.984346   \n",
       "258409     004dd1a01106608a  chr10:100339340-100339978          3.327292   \n",
       "258410     004dd1a01106608a  chr10:101066169-101066970          3.590057   \n",
       "...                     ...                        ...               ...   \n",
       "216251363  ffffeba9dbf2bab1     chrY:19499847-19500746          3.342186   \n",
       "216251364  ffffeba9dbf2bab1     chrY:19566990-19567754          1.154666   \n",
       "216251365  ffffeba9dbf2bab1     chrY:20735890-20736626          7.049924   \n",
       "216251366  ffffeba9dbf2bab1     chrY:21260054-21260842          3.967924   \n",
       "216251367  ffffeba9dbf2bab1       chrY:7273431-7274301          1.250685   \n",
       "\n",
       "          chromosome      start  \n",
       "258406         chr10  100006075  \n",
       "258407         chr10  100229082  \n",
       "258408         chr10  100286187  \n",
       "258409         chr10  100339340  \n",
       "258410         chr10  101066169  \n",
       "...              ...        ...  \n",
       "216251363       chrY   19499847  \n",
       "216251364       chrY   19566990  \n",
       "216251365       chrY   20735890  \n",
       "216251366       chrY   21260054  \n",
       "216251367       chrY    7273431  \n",
       "\n",
       "[215992962 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_end(all_str):\n",
    "    return all_str[all_str.find(\"-\")+1:] \n",
    "\n",
    "\n",
    "atac[\"end\"] = atac[\"location\"].apply(get_end)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "kaggle_scp",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
